krun: implement support for passt networking

slp opened the pull request:
Introduce a new configuration flag (krun.use_passt) that enables users to request the microVM to use passt based networking instead of TSI.

Right now, the container needs to configure the network by itself, but the next version of libkrun will embed a DHCP client in init.c

slp committed:
handlers: add new hook to close fds

Some handlers may need to preserve more fds than those related to
stdio. Let's allow them to provide their own hook for doing that.

Signed-off-by: Sergio Lopez <slp@redhat.com>
Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

slp committed:
krun: implement support for passt networking

Introduce a new configuration flag (krun.use_passt) that enables users
to request the microVM to use passt based networking instead of TSI.

Right now, the container needs to configure the network by itself, but
the next version of libkrun will embed a DHCP client in init.c

Signed-off-by: Sergio Lopez <slp@redhat.com>
Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

sbrivio-rh commented on src/libcrun/handlers/krun.c:
```diff
@@ -464,6 +477,69 @@ libkrun_exec (void *cookie, libcrun_container_t *container, const char *pathname
+  else
+    {
+      /* We need to make sure passt has already started before continuing. A
+         simple way to do it is with a blocking read on its stdout. */
```
Actually, you can skip this whole dance by *not* passing `-f`: passt behaves like usual UNIX daemons and forks into background when it's ready, and if you give `-1` / `--one-off`, it terminates by itself once the guest disconnects.

Conversely, getting something from its stdout doesn't guarantee that it's ready. Which is the only (minor) issue I spotted here, by the way, so if you need to stick to this approach for whatever reason it looks fine to me as well.

slp replied:
Indeed, this looks much better, thanks for the suggestion!

slp commented:
> Maybe an approach similar to https://github.com/AsahiLinux/muvm/pull/111 (tiny DHCP client embedded in muvm) would make sense here as well.

That's exactly what I was planning to do, basically by translating your Rust implementation into C for `init/init.c`.

slp committed:
krun: process the vm configuration earlier

Some decisions, such as whether we need to start passt or not, need to
happen before we switch to the container's mount namespace. With this
change we move processing the VM configuration to the
HANDLER_CONFIGURE_BEFORE_USERNS phase, so we can have the information
early enough.

Signed-off-by: Sergio Lopez <slp@redhat.com>
Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

slp committed:
krun: consolidate configuration in configure_vm

Since the introduction of libkrun_configure_vm, the legacy path was
only walked on very exceptional situations. Let's consolidate both
configuration mechanisms in libkrun_configure_vm.

While there, make GPU a configurable option, instead of relying in
heuristics.

Signed-off-by: Sergio Lopez <slp@redhat.com>
Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

slp committed:
krun: ignore RAM configurations below 128MB

The minimum viable amount for RAM for a microVM is 128MB, ignore
configurations asking for less than that.

Signed-off-by: Sergio Lopez <slp@redhat.com>
Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

slp commented:
> docker:
>     Error response from daemon:
>     failed to create task for container: 
>     failed to create shim task:
>     OCI runtime create failed:
>     failed configuring mounts for handler at phase:
>     HANDLER_CONFIGURE_AFTER_MOUNTS:
>     start passt:
>     File exists: unknown.

I'm not sure what's leading to EEXIST, could you please trace it with `strace`?

tylerfanelli commented:
@slp @sbrivio-rh With the following diff, passt is able to work with nitro:
```
diff --git a/src/libcrun/handlers/krun.c b/src/libcrun/handlers/krun.c
index a2555cf5..54006b1a 100644
--- a/src/libcrun/handlers/krun.c
+++ b/src/libcrun/handlers/krun.c
@@ -545,11 +545,8 @@ libkrun_start_passt (void *cookie, libcrun_container_t *container)
 
   char *const argv[] = {
     (char *) "passt",
-    (char *) "--no-dhcp-dns",
     (char *) "-t",
     (char *) "all",
-    (char *) "-u",
-    (char *) "all",
     (char *) "--fd",
     fd_as_str,
     NULL
```
For now, I can add a commit to this branch to only add these options on non-nitro, but eventually I'll need to add a DHCP client (likely built from melding https://github.com/containers/libkrun/pull/589 into aws-nitro) so everything works smoothly.

tylerfanelli committed:
krun: Build passt argv based on nitro variant

The krun-awsnitro variant requires the removal of a few command line
arguments to bring passt networking support.

This is a temporary solution intended until a proper DHCP client can be
added to krun-awsnitro.

Signed-off-by: Tyler Fanelli <tfanelli@redhat.com>
Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

slp commented:
@giuseppe PTAL

giuseppe commented:
could you please rebase on top of main?  Do we need the merge commit or can it be rebased too?

giuseppe commented:
I've pushed to your branch, I'd like to get this in time for the new release

giuseppe committed:
.codespellrc: ignore passt

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

giuseppe reviewed (approved):
LGTM

siteshwar commented on src/libcrun/handlers/krun.c:
```diff
@@ -233,20 +245,36 @@ libkrun_parse_resource_configuration (yajl_val *config_tree, libcrun_container_t
+        ram_mib = LIBKRUN_DEFAULT_RAM_MIB;
+    }
 
   krun_set_vm_config = dlsym (handle, "krun_set_vm_config");
```
OpenScanHub reports this variable is [not used](https://openscanhub.fedoraproject.org/task/109914/log/added.html#def3) anymore.

giuseppe replied:
but it is used at line 284?

siteshwar replied:
Sorry! The reference is to a [different line](https://github.com/slp/crun/blob/acb904c6abc43f3e71d6e1abe73b1dd9bb035f7d/src/libcrun/handlers/krun.c#L412).

giuseppe replied:
thanks!

Added a patch to https://github.com/containers/crun/pull/2071 to fix it
